sway/desktop/layer_shell.c

```diff
@@ -0,0 +1,317 @@
+			&usable_area, true);
+	memcpy(&output->usable_area, &usable_area, sizeof(struct wlr_box));
+
+	if (memcmp(&usable_area_before,
+			&usable_area, sizeof(struct wlr_box)) != 0) {
+		wlr_log(L_DEBUG, "arrange");
+		arrange_windows(output->swayc, -1, -1);
+	}
+
+	// Arrange non-exlusive surfaces from top->bottom
+	usable_area.x = usable_area.y = 0;
+	wlr_output_effective_resolution(output->wlr_output,
+			&usable_area.width, &usable_area.height);
+	arrange_layer(output, &output->layers[ZWLR_LAYER_SHELL_V1_LAYER_OVERLAY],
+			&usable_area, false);
+	arrange_layer(output, &output->layers[ZWLR_LAYER_SHELL_V1_LAYER_TOP],
+			&usable_area, false);
+	arrange_layer(output, &output->layers[ZWLR_LAYER_SHELL_V1_LAYER_BOTTOM],
+			&usable_area, false);
+	arrange_layer(output, &output->layers[ZWLR_LAYER_SHELL_V1_LAYER_BACKGROUND],
+			&usable_area, false);
+}
+
+static void handle_output_destroy(struct wl_listener *listener, void *data) {
+	struct sway_layer_surface *sway_layer =
+		wl_container_of(listener, sway_layer, output_destroy);
+	sway_layer->layer_surface->output = NULL;
+	wl_list_remove(&sway_layer->output_destroy.link);
+	wl_list_remove(&sway_layer->output_mode.link);
+	wlr_layer_surface_close(sway_layer->layer_surface);
+}
+
+static void handle_output_mode(struct wl_listener *listener, void *data) {
+	struct wlr_output *output = data;
+	arrange_layers((struct sway_output *)output->data);
+}
+
+static void handle_output_transform(struct wl_listener *listener, void *data) {
+	struct wlr_output *output = data;
+	arrange_layers((struct sway_output *)output->data);
+}
+
+static void handle_surface_commit(struct wl_listener *listener, void *data) {
+	struct sway_layer_surface *layer =
+		wl_container_of(listener, layer, surface_commit);
+	struct wlr_layer_surface *layer_surface = layer->layer_surface;
+	struct wlr_output *wlr_output = layer_surface->output;
+	if (wlr_output != NULL) {
+		struct sway_output *output = wlr_output->data;
+		struct wlr_box old_geo = layer->geo;
+		arrange_layers(output);
+		if (memcmp(&old_geo, &layer->geo, sizeof(struct wlr_box)) != 0) {
+			// TODO DAMAGE apply whole surface from previous and new geos
+		} else {
+			// TODO DAMAGE from surface damage
+		}
+	}
+}
+
+static void unmap(struct wlr_layer_surface *layer_surface) {
+	// TODO DAMAGE
+}
+
+static void handle_destroy(struct wl_listener *listener, void *data) {
+	struct sway_layer_surface *sway_layer = wl_container_of(
+			listener, sway_layer, destroy);
+	if (sway_layer->layer_surface->mapped) {
+		unmap(sway_layer->layer_surface);
+	}
+	wl_list_remove(&sway_layer->link);
+	wl_list_remove(&sway_layer->destroy.link);
+	wl_list_remove(&sway_layer->map.link);
+	wl_list_remove(&sway_layer->unmap.link);
+	wl_list_remove(&sway_layer->surface_commit.link);
+	wl_list_remove(&sway_layer->output_destroy.link);
+	wl_list_remove(&sway_layer->output_mode.link);
+	wl_list_remove(&sway_layer->output_transform.link);
```

Comment:
Maybe listeners have been removed in `handle_output_destroy` already. Should guard this with a 

```c
if (sway_layer->output != NULL)
```